(code bounty) The tram is now unstoppably powerful. it cannot be stopped, it cannot be slowed, it cannot be reasoned with. YOU HAVE NO IDEA HOW READY YOU ARE

Kylerace opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->
ever see the tram take 10 milliseconds per movement to move 2100 objects? now you have
https://user-images.githubusercontent.com/15794172/166198184-8bab93bd-f584-4269-9ed1-6aee746f8f3c.mp4

## About The Pull Request
fixes #66887

done for the code bounty posted by @MMMiracles to optimize the tram so that it can be sped up. the tram is now twice as fast, firing every tick instead of every 2 ticks. and is now around 10x cheaper to move. also adds support for multiz trams, as in trams that span multiple z levels.

the tram on master takes around 10-15 milliseconds per movement with nothing on it other than its starting contents. why is this? because the tram is the canary in the coal mines when it comes to movement code, which is normally expensive as fuck. the tram does way more work than it needs to, and even finds new ways to slow the game down. I'll walk you through a few of the dumber things the tram currently does and how i fixed them. 

1. the tram, at absolute minimum, has to move 55 separate industrial_lift platforms once per movement. this means that the tram has to unregister its entered/exited signals 55 times when "the tram" as a singular object is only entering 5 new turfs and exiting 5 old turfs every movement, this means that each of the 55 platforms calculates their own destination turfs and checks their contents every movement. The biggest single optimization in this pr was that I made the tram into a single 5x11 multitile object and made it only do entering/exiting checks on the 5 new and 5 old turfs in each movement. 
2. way too many of the default tram contents are expensive to move for something that has to move a lot. fun fact, did you know that the walls on the tram have opacity? do you know what opacity does for movables? it makes them recalculate static lighting every time they move. did you know that the tram, this entire time, was taking JUST as much time spamming SSlighting updates as it was spending time in SStramprocess? well it is! now it doesnt do that, the walls are transparent. also, every window and every grille on the tram had the atmos_sensitive element applied to them which then added connect_loc to them, causing them to update signals every movement. that is also dumb and i got rid of that with snowflake overrides. Now we must take care to not add things that sneakily register to Moved() or the moved signal to the roundstart tram, because that is dumb, and the relative utility of simulating objects that should normally shatter due to heat and conduct heat from the atmosphere is far less than the cost of moving them, for this one object.
3. all tram contents physically Entered() and Exited() their destination and old turfs every movement, even though because they are on a tram they literally do not interact with the turf, the tram does. also, any objects that use connect_loc or connect_loc behalf that are on the same point on the tram also interact with each other because of this. now all contents of the tram act as if theyre being abstract_move()'d to their destination so that (almost) nothing thats in the destination turf or the exit turf can react to the event of "something laying on the tram is moving over you". the rare things that DO need to know what is physically entering or exiting their turf regardless of whether theyre interacting with the ground can register to the abstract entered and exited signals which are now always sent.
4. many of the things hooked into Moved(), whether it be overrides of Moved() itself, or handlers for the moved signal, add up to a LOT of processing time. especially for humans. now ive gotten rid of a lot of it, mostly for the tram but also for normal movement. i made footsteps (a significant portion of human movement cost) not do any work if the human themselves didnt do the movement. i optimized has_gravity() a fair amount, and then realized that since everything on the tram isnt changing momentum, i didnt actually need to check gravity for the purposes of drifting (newtonian_move() was taking a significant portion of the cost of movement at some points along the development process). so now it simply doesnt call newtonian_move() for movements that dont represent a change in momentum (by default all movements do). 

also i put effort into 1. better organizing tram/lift code so that most of it is inside of a dedicated modules folder instead of scattered around 5 generic folders and 2. moved a lot of behavior from lift platforms themselves into their lift_master_datum since ideally the platforms would just handle moving themselves, while any behavior involving the entire lift such as "move to destination" and "blow up" would be handled by the lift_master_datum.

also 
https://user-images.githubusercontent.com/15794172/166220129-ff2ea344-442f-4e3e-94f0-ec58ab438563.mp4
multiz tram (this just adds the capability to map it like this, no tram does this)

## Actual Performance Differences
to benchmark this, i added a world.Profile(PROFILER_START) and world.Profile(PROFILER_STOP) to the tram moving, so that it generates a profiler output of all tram movement without any unrelated procs being recorded (except for world.Profile() overhead). this made it a lot easier to quantify what was slowing down both the tram and movement in general. and i did 3 types of tests on both master and my branch. 

also i should note that i sped up the "master"  tram test to move once per tick as well, simply because the normal movement speed seems unbearably slow now. so all recorded videos are done at twice the speed of the real tram on master. this doesnt affect the main thing i was trying to measure: cost for each movement. 

the first test was the base tram, containing only my player mob and the movables starting on the tram roundstart. on master, this takes around 13 milliseconds or so on my computer (which is pretty close to what it takes on the servers), on this branch, it takes between 0.9-1.3 milliseconds.

ALSO in these benchmarks youll see that tram/proc/travel() will vary significantly between the master and optimized branches. this is 100% because there are 55 times more platforms moving on master compared to the master branch, and thus 55x more calls to this proc. every test was recorded with the exact same amount of distance moved

here are the master and optimized benchmark text files:
master
[master base tram.txt](https://github.com/tgstation/tgstation/files/8601662/master.base.tram.txt)
https://user-images.githubusercontent.com/15794172/166210149-f118683d-6f6d-4dfb-b9e4-14f17b26aad8.mp4
also this shows the increased SSlighting usage resulting from the tram on master spamming updates, which doesnt happen on the optimized branch

optimized
[optimization base tram.txt](https://github.com/tgstation/tgstation/files/8601664/optimization.base.tram.txt)
https://user-images.githubusercontent.com/15794172/166206280-cd849aaa-ed3b-4e2f-b741-b8a5726091a9.mp4

the second test is meant to benchmark the best case scaling cost of moving objects, where nothing extra is registered to movement besides the bare minimum stuff on the /atom/movable level. Each of the open tiles of the tram had 1 bluespace rped filled with parts dumped onto it, to the point that the tram in total was moving 2100 objects. the vast majority of these objects did nothing special in movement so they serve as a good base case. only slightly off due to the rped's registering to movement. 

on master, this test takes over 100 milliseconds per movement
[master 2000 obj's.txt](https://github.com/tgstation/tgstation/files/8601733/master.2000.obj.s.txt)
https://user-images.githubusercontent.com/15794172/166210560-f4de620d-7dc6-4dbd-8b61-4a48149af707.mp4

when optimized, about 10 milliseconds per movement
https://user-images.githubusercontent.com/15794172/166208654-bc10086b-bbfc-49fa-9987-d7558109cc1d.mp4
[optimization 2000 obj's.txt](https://github.com/tgstation/tgstation/files/8601734/optimization.2000.obj.s.txt)

the third test is 300 humans spawned onto the tram, meant to test all the shit added on to movement cost for humans/carbons. in retrospect this test is actually way too biased in favor of my optimizations since the humans are all in only 3 tiles, so all 100 humans on a tile are reacting to the other 99 humans movements, which wouldnt be as bad if they were distributed across 20 tiles like in the second test. so dont read into this one too hard. 

on master, this test takes 200 milliseconds
[master 300 catgirls.txt](https://github.com/tgstation/tgstation/files/8601769/master.300.catgirls.txt)

when optimized, this takes about 13-14 milliseconds.
[optimization 300 catgirls on ram ranch.txt](https://github.com/tgstation/tgstation/files/8601775/optimization.300.catgirls.on.ram.ranch.txt)

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
the tram is literally 10x cheaper to move. and the code is better organized. 
currently on master the tram is as fast as running speed, meaning it has no real relative utility compared to just running the tracks (except for the added safety of not having to risk being ran over by the tram). now the tram of which we have an entire map based around can be used to its full potential.

also, has some fixes to things on the tram reacting to movement. for example on master if you are standing on a tram tile that contains a banana and the TRAM moves, you will slip if the banana was in that spot before you (not if you were there first however). this is because the banana has no concept of relative movement, you and it are in the same reference frame but the banana, which failed highschool physics, believes you to have moved onto it and thus subjected you to the humiliation of an unjust slipping. now since tram contents that dont register to abstract entered/exited cannot know about other tram contents on the same tile during a movement, this cannot happen. 

also, you no longer make footstep sounds when the tram moves you over a floor

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## TODO
mainly opened it now so i can create a stopping point and attend to my other now staling prs, we're at a state of functionality far enough to start testmerging it anyways.

- [x] add a better way for admins to be notified of the tram overloading the server if someone purposefully stuffs it with as much shit as they can, and for admins to clear said shit.
- [x] automatically slow down the tram if SStramprocess takes over like, 10 milliseconds complete. the tram still cant really check tick and yield without introducing logic holes, so making sure it doesnt take half of the tick every tick is important
- [x] go over my code to catch dumb shit i forgot about, there always is for these kinds of refactors because im very messy
- [x] remove the area based forced_gravity optimization its not worth figuring out why it doesnt work
- [x] fix the inevitable merge conflict with master lol
- [x] create an icon for the tram_tunnel area type i made so that objects on the tram dont have to enter and exit areas twice in a cross-station traversal
- [x] add an easy way to vv tram lethality for mobs/things being hit by it. its an easy target in another thing i already wanted to do: a reinforced concept of shared variables from any particular tram platform and the entire tram itself. admins should be able to slow down the tram by vv'ing one platform and have it apply to the entire tram for example.

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
balance: the tram is now twice as fast, pray it doesnt get any faster (it cant without raising world fps)
performance: the tram is now about 10 times cheaper to move for the server
add: mappers can now create trams with multiple z levels
code: industrial_lift's now have more of their behavior pertaining to "the entire lift" being handled by their lift_master_datum as opposed to belonging to a random platform on the lift. 
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Kylerace committed:
makes the fov_image actually work

Kylerace committed:
basic unregistering signals before and after movement

Kylerace committed:
adds industrial_lift/proc/group_move() for better efficiency

Kylerace committed:
manually edits tram objects to have 0 opacity

turns out moving the tram caused SSLighting to have a fuckton of usage because 14 objects on the tram had opacity = 1 which caused them to rebuild lighting every fucking movement. so now they dont do that

Kylerace committed:
optimizes has_gravity()

Kylerace committed:
removes opacity from all tram base objects to stop recalculating lights

Kylerace committed:
makes only border turfs check for collisions and other optimizations

Kylerace committed:
makes doMove() work with multitile movement

Kylerace committed:
makes the tram move as one multitile platform (doesnt work rn)

Kylerace committed:
fixes a lot of multitile tram platform bugs with moving

now it doesnt always break at the end of hte movement. but it still doesnt work

Kylerace committed:
moves the tram landmarks to the bottom left of where the tram goes

works for a multiz tram platform because loc is the lower left portion

Kylerace committed:
multitile tram now MOSTLY works and its fast

Kylerace committed:
makes the tram computer not have light to avoid update spam

Kylerace committed:
as anything in a loop

Kylerace committed:
misc improvements to has_gravity()

Kylerace committed:
optimizes has_gravity() more

Kylerace committed:
optimizes group_move() and travel() for the tram

Kylerace committed:
more Moved() optimizations

Kylerace committed:
fixes footsteps happening when being moved by the tram

Kylerace committed:
more optimizations to tram movement, it cant get much better

adds behavior to only affect the glide_size of the movables in lift_load that need it, further microopts to travel()

Kylerace committed:
gets rid of move_stacks and makes newtonian_move() only call if alone

Kylerace committed:
makes areas add forced gravity on enter if they force gravity

Kylerace committed:
moves updating lights and parallax updating to the Moved signal

Kylerace committed:
adds tram ovverrides for windoors that dont have atmos_sensitive

Kylerace committed:
makes atmos_sensitive actually undo Moved registrations

Kylerace committed:
makes bloodysoles and overlay_lighting register to movement

Kylerace committed:
gets rid of the second Moved() loop in group_move()

Kylerace committed:
sets whole tram tunnel to 1 area and fixes more performance

Kylerace committed:
cherry picks frozen element from other branch

now we dont have to check for frozen objects in /turf/open/Entered()

Kylerace committed:
cherry picks open storage containers in important_recursive_contents

this speeds up moving while having an storage container opened since now if you just have your backpack opened we just need to check if the backpack is inside your important_recursive_contents list

Kylerace committed:
makes the cherry-pick work

Kylerace committed:
cherry picks sound optimization from old branch

Kylerace committed:
cherry picks another old commit optimizing sound

Kylerace committed:
cherry pick closeby_client_mobs from old branch

Kylerace committed:
caches linked z levels and more ops to movement

Kylerace committed:
removes parallax updating via signals for now

Kylerace committed:
adds an element to handle human interactions when on fire

Kylerace committed:
makes update_turf_movespeed() do nothing if no changes happened

Kylerace committed:
slight optimizations to client/Move

Kylerace committed:
fixes registrations for overlay_lighting component

Kylerace committed:
kills TODOKYLER

Kylerace committed:
moves the "central" industrial lift to the lower left corner

Kylerace committed:
gets rid of abstract_entered() and abstract_exited() for just signals

Kylerace committed:
makes hotspots use the abstract_entered signal

Kylerace committed:
better slowdown logic and puts definitions where they belong

Kylerace committed:
puts force_no_gravity defs where they belong

Kylerace committed:
gets rid of dumb !inertia_moving check for has_gravity()

Kylerace committed:
makes the multitile tram look like separate platforms, misc cleanup

Kylerace committed:
adds more early return checks to playsound()

Kylerace committed:
kills TODOKYLER

Kylerace committed:
makes SStramprocess wait 0.5

Kylerace committed:
fixes z level gravity for z levels with no generators but traits

Kylerace committed:
fixes shuttles not working due to docking ports drifting

Kylerace committed:
docs shuttle code

Kylerace committed:
removes move_stacks

Kylerace committed:
organizes industrial_lift code into several files in a modules folder

Kylerace committed:
makes multiz trams possible

Kylerace committed:
removes parallax changes

Kylerace committed:
makes a tram subtype of lift_masterand moves tram controls to the module

Kylerace committed:
adds better lift linking, moves behavior to lift_masters, 

it technically compiles but needs a lot more work. ideally i can have all tramwide behavior be moved to lift_master datums instead of the platforms, which will handle only moving themselves

Kylerace committed:
adds ID landmark to tram

Kylerace committed:
removes closeby_client_mobs because it doesnt help

Kylerace committed:
docs, minor fixes, reverts failed changes, kills TODOKYLER's

Kylerace committed:
kills the profiling stuff

Kylerace commented on code/controllers/subsystem/explosions.dm:
```diff
@@ -528,18 +528,18 @@ SUBSYSTEM_DEF(explosions)
 
 		if(distance <= round(near_distance + world.view - 2, 1)) // If you are close enough to see the effects of the explosion first-hand (ignoring walls)
-			listener.playsound_local(epicenter, null, 100, TRUE, frequency, S = near_sound)
+			listener.playsound_local(epicenter, null, 100, TRUE, frequency, sound_to_use = near_sound)
```
byproduct of a cherry pick, revert

Kylerace replied:
eh actually whatever S is dumb

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -290,10 +243,6 @@ SUBSYSTEM_DEF(spatial_grid)
 
-	//cache for sanic speeds
-	var/cells_on_y_axis = src.cells_on_y_axis
-	var/cells_on_x_axis = src.cells_on_x_axis
```
dumb, revert

Kylerace replied:
wait no this change isnt dumb, each of these datum vars are only accessed once per call so caching it to a proc var makes no sense

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -319,6 +268,36 @@ SUBSYSTEM_DEF(spatial_grid)
+///returns a square of grid cells outward from the spatial grid cell encompassing the center atom.
+///num_cells_outward_of_center is the number of cells between the center cell and the outer layer. making this 0 just returns the grid cell at the center.
+///use this if you want to specify range in grid cells not turfs.
+/datum/controller/subsystem/spatial_grid/proc/get_block_of_cells(atom/center, num_cells_outward_of_center)
```
byproduct of cherry pick, revert

Kylerace commented on code/datums/components/bloodysoles.dm:
```diff
@@ -53,19 +53,42 @@
 	parent_item.update_slot_icon()
 
+
+/datum/component/bloodysoles/proc/reset_bloody_shoes()
```
bloody_shoes previously registered to every movement even when it only did anything on a normal movement when said shoes were covered in blood. in order to lower human/carbon movement cost i made it only register to the moved signal if said shoes are actually bloody

Kylerace commented on code/datums/components/overlay_lighting.dm:
```diff
@@ -245,8 +245,9 @@
-		RegisterSignal(new_holder, COMSIG_MOVABLE_MOVED, .proc/on_holder_moved)
 		RegisterSignal(new_holder, COMSIG_LIGHT_EATER_QUEUE, .proc/on_light_eater)
+		if(overlay_lighting_flags & LIGHTING_ON)
+			RegisterSignal(new_holder, COMSIG_MOVABLE_MOVED, .proc/on_holder_moved)
```
parent must always have its movement registered, but the holder (which is usually the human mob, since pda's carry flashlights) should only have their movement registered if the light is actually on and needs updates

Kylerace commented on code/datums/components/storage/storage.dm:
```diff
@@ -420,6 +441,9 @@ GLOBAL_LIST_EMPTY(cached_storage_typecaches)
 	M.set_active_storage(src)
+	if(ismovable(real_location))
+		var/atom/movable/movable_loc = real_location
+		movable_loc.become_active_storage(src)
```
fun fact: if you open your backpack and move, your mob then performs a BFS to find out whether its currently active storage object is still within reach. now i added active storage objects to important_recursive_contents so that if said object is within your inventory you early out of the BFS performed by CanReach()

Kylerace commented on code/datums/elements/atmos_sensitive.dm:
```diff
@@ -22,6 +22,7 @@
 /datum/element/atmos_sensitive/Detach(datum/source)
 	var/atom/us = source
 	us.RemoveElement(/datum/element/connect_loc, pass_on)
+	UnregisterSignal(source, COMSIG_MOVABLE_MOVED)
```
it wasnt doing this

LemonInTheDark replied:
😳

Kylerace commented on code/datums/elements/footstep.dm:
```diff
@@ -87,7 +87,7 @@
 		return
 
-	if(steps != 0 && !source.has_gravity(turf)) // don't need to step as often when you hop around
+	if(steps != 0 && source.has_gravity()) // don't need to step as often when you hop around
```
wtf am i doing fix this

Kylerace commented on code/datums/elements/footstep.dm:
```diff
@@ -117,10 +117,10 @@
 	SIGNAL_HANDLER
 
-	if (SHOULD_DISABLE_FOOTSTEPS(source))
+	if (SHOULD_DISABLE_FOOTSTEPS(source) || !momentum_change)
```
approximation of "you didnt actually move with your legs since something else moved you, so dont make footstep sounds"

LemonInTheDark replied:
should comment this

Ghilker commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -114,7 +114,7 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 		TOTAL_MOLES(cached_gases, .)
 		. *= R_IDEAL_GAS_EQUATION * temperature / volume
-		return
+		return .
```
Wha

Mothblocks replied:
https://github.com/tgstation/tgstation/blob/master/.github/guides/STANDARDS.md#dot-variable-

It's correct

Kylerace commented on code/datums/elements/footstep.dm:
```diff
@@ -134,21 +134,31 @@
+				bare_footstep_sounds[source_loc.barefootstep][3] + e_range + range_adjustment, falloff_distance = 1, vary = sound_vary)
+
+	if(heard_clients)
+		play_fov_effect(source, 5, "footstep", direction, ignore_self = TRUE, override_list = heard_clients)
```
playsound() now returns the actual listeners that head the message. now whenever we footstep we dont have to call get_hearers_in_view() to find fov see-ers

LemonInTheDark replied:
chad

Kylerace commented on code/datums/elements/forced_gravity.dm:
```diff
@@ -1,16 +1,21 @@
+	///whether we will override the turf if it forces no gravity
+	var/ignore_turf_gravity
+	///whether this forced gravity instance is from an area that adds gravity
+	var/area_based
```
revert, doesnt work and isnt worth fixing

Kylerace commented on code/datums/elements/frozen.dm:
```diff
@@ -0,0 +1,47 @@
+/datum/element/frozen
+	element_flags = ELEMENT_DETACH
+
+/datum/element/frozen/Attach(datum/target)
```
previously checking if an object was frozen and updating it was done in /turf/open/Entered() but frozen objects are a tiny minority of movables entering turfs. so it makes sense to offload this into an element that does the work we need

Kylerace commented on code/datums/elements/on_fire_connector.dm:
```diff
@@ -0,0 +1,48 @@
+///simple element to handle: 1. mobs that are on fire bumping into other mobs
+///and 2. mobs on fire moving into a turf containing mobs not on fire, but not bumping into them
+/datum/element/on_fire_connector
```
@LemonInTheDark  one of the two connect_loc registers to normal human movement. it was the less ugly of the two to move away from connect_loc, so i did this one.

Kylerace commented on code/game/area/areas.dm:
```diff
@@ -417,7 +417,10 @@ GLOBAL_LIST_EMPTY(teleportlocs)
 	SEND_SIGNAL(src, COMSIG_AREA_ENTERED, arrived, old_area)
-	if(!LAZYACCESS(arrived.important_recursive_contents, RECURSIVE_CONTENTS_AREA_SENSITIVE))
+	if(has_gravity)
+		arrived.AddElement(/datum/element/forced_gravity, has_gravity, FALSE, TRUE)
```
bad change, revert

Kylerace commented on code/game/atoms_movable.dm:
```diff
@@ -3,8 +3,6 @@
 	appearance_flags = TILE_BOUND|PIXEL_SCALE
 
-	///how many times a this movable had movement procs called on it since Moved() was last called
-	var/move_stacks = 0
```
move_stacks should no longer be needed since connect_loc_behalf is no longer eating us alive. but we'll see, might revert

LemonInTheDark replied:
didn't we need this to ensure signals were sent at the right time?

Kylerace commented on code/game/atoms_movable.dm:
```diff
@@ -588,8 +586,6 @@
 				if(!. && set_dir_on_move)
 					setDir(first_step_dir)
-				else if (!inertia_moving)
-					newtonian_move(direct)
```
already done by Moved()

LemonInTheDark replied:
no, it's not, because moving_diagonally is set during that move, at least for the parallax

Kylerace commented on code/game/atoms_movable.dm:
```diff
@@ -879,12 +891,13 @@
 	. = FALSE
-	move_stacks++
 	var/atom/oldloc = loc
+	var/is_multi_tile = bound_width > 32 || bound_height > 32
```
doMove() wasnt multitile friendly, now it is

Kylerace replied:
I should also make abstract_move() work with multitile movement

Kylerace commented on code/game/sound.dm:
```diff
@@ -23,120 +23,102 @@
-	if(!ignore_walls) //these sounds don't carry through walls
-		listeners = listeners & hearers(maxdistance,turf_source)
+	if(ignore_walls)
+		listeners = SSspatial_grid.orthogonal_range_search(turf_source, SPATIAL_GRID_CONTENTS_TYPE_CLIENTS, maxdistance)
```
this was another cherry pick from a scrapped movement optimization pr and eh honestly i wasnt able to find any advantage to using the grid over iterating clients in the z level surprisingly enough, only thing i need here is returning real hearers

Kylerace commented on code/game/sound.dm:
```diff
@@ -149,35 +131,35 @@ distance_multiplier - Can be used to multiply the distance at which the sound is
+			sound_to_use.echo[4] = 0 //RoomHF setting, 0 means normal reverb.
 
-	SEND_SOUND(src, S)
+	SEND_SOUND(src, sound_to_use)
```
fun fact this is like a third of playsound()'s total cost for some reason

Kylerace replied:
correction, about 12%

Kylerace commented on code/modules/atmospherics/environmental/LINDA_fire.dm:
```diff
@@ -98,6 +98,7 @@
 	air_update_turf(FALSE, FALSE)
 	var/static/list/loc_connections = list(
 		COMSIG_ATOM_ENTERED = .proc/on_entered,
+		COMSIG_ATOM_ABSTRACT_ENTERED = .proc/on_entered,
```
even with the tram we are still entering the bounds of the turf and thus touch the air, so hotspots should register to things entering

LemonInTheDark replied:
?

MMMiracles commented:
unfathomably based

Palpetinus commented:
>currently on master the tram is as fast as running speed, meaning it has no real relative utility compared to just running the tracks (except for the added safety of not having to risk being ran over by the tram)

The coders invented a problem solely to then sell us the cure, don't fall for their tricks!

In all seriousness, this is amazingly based. And just as I was working on a circuit that repeatedly fires the West and East destinations to turn the tram into a death-machine, impeccable timing.

Shadowflame909 commented:
How fast does it have to go for my sprite to phase onto the tram before I get shuttle gibbed

timothymtorres commented:
The multi-z tram should be a thing on tramstation.  It looks super cool.

lemshake commented:
finally a good pull request

MrDoomBringer commented:
pr of the year

Zonespace27 commented:
Doing god's work.

Maurukas commented:
> also adds support for multiz trams, as in trams that span multiple z levels.

If this does not result in at least one subsequent PR with the FUCK tag I am going to be disappointed. I understand a fraction of what this is changing, and even that is impressive as hell.

Kylerace committed:
whoops i broke footsteps

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into fasterer_tram

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -114,7 +114,7 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 		var/cached_gases = gases
 		TOTAL_MOLES(cached_gases, .)
 		. *= R_IDEAL_GAS_EQUATION * temperature / volume
-		return
+		return .
```
For the record this is very very very very technically slower. It's alright here though.
That said, this syntax still makes me want to barf. do this instead
```
return . * R_IDEAL_GAS_EQUATION * temperature / volume 
```

LemonInTheDark replied:
also something something atmos code is bad on purpose

ATH1909 commented:
https://youtu.be/wZwQvGvZ86k?t=27

Kylerace committed:
fixes merge with master and allows multiple trams

Kylerace committed:
kills dumb area gravity change

Kylerace committed:
slows down the tram if its too fast and does mulitile trams centrally

Kylerace committed:
fixes dumb merge conflict artifacts in the tram tunnel

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into fasterer_tram

Kylerace committed:
wtf

Kylerace committed:
Hilbert->hilbert

Kylerace committed:
fixes stacking move registrations in turn_on()

Kylerace committed:
adds industrial_lift's to exemptions for create_and_destroy

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into fasterer_tram

Kylerace committed:
i cannot be assed to make this work AGAIN, reverts fire optimization

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into fasterer_tram

Kylerace committed:
resets tramstatiopn.dmm to master because fuck conflicts

Kylerace committed:
why the fuck were tramwalls given opacity again

Kylerace committed:
remaps in tram stuff

Kylerace committed:
fixes linked_z_levels list not expanding when adding new z levels

tralezab commented:
conflict! but blessed

tralezab commented:
that makes me so mad it can't be both

Kylerace committed:
merge master and fixes conflicts

Kylerace committed:
why did i keep the old file from master

Rocko1345 commented:
Is there any way to make the tram more capable of causing wounds on impact?

Kylerace committed:
maps in a lower z level to the tram to test that feature

private-tristan commented:
Tram now seems to leave behind shutters built on it (previously the shutters were left on the tram, and able to be toggled on/off to make a tramfort) please fix so that I can tramfort again!

https://user-images.githubusercontent.com/54422837/167279386-d5f942ff-1e22-4b6d-865f-0bd5644a06c5.mp4

Kylerace committed:
makes the tram not double up lift platforms

Kylerace committed:
adds error condition and removes commented code

Kylerace committed:
makes the tram error if there are >1 platforms on a tile

Kylerace committed:
makes the slowdown process more robust

Kylerace committed:
readds lift id icon state

Kylerace commented:
@Rocko1345 
an admin can vv the tram's lift_master_datum's collision_lethality var to be higher.

Kylerace committed:
improves slowdown logic

Kylerace committed:
makes it possible to reset the tram to its initial state

Kylerace committed:
makes the tram add newly created objects to lift_load again

Kylerace committed:
merges master and fixes conflict

Kylerace commented:
@private-tristan should be fixed now

Kylerace commented:
admins can now reset the tram to its original state by calling reset_lift_contents() on the trams lift_master_datum and can have some limited control over what gets deleted

Kylerace committed:
removes central trams because they arent needed anymore

Kylerace committed:
merges master and fixes conflict

Kylerace committed:
fucking manual conflict resolution

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into fasterer_tram

Kylerace commented:
btw im testing the multiz tram feature for now

Kylerace committed:
fixes camera mobs being included in the tram's lift_load

Kylerace committed:
removes the lower tram it is tested now

Kylerace committed:
gets rid of cherry pick artifact and does a lemon review

Kylerace committed:
fixes frozen element applying forever

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into fasterer_tram

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into fasterer_tram

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into fasterer_tram

Kylerace committed:
fixes the crossing signals for the tram

Kylerace committed:
merges master

Kylerace committed:
Merge remote-tracking branch 'origin/fasterer_tram' into fasterer_tram

Kylerace committed:
merges master

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into fasterer_tram

Kylerace committed:
whoops i fucked up that merge, fixed

MMMiracles commented:
tram my beloved...

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into fasterer_tram
fixes conflict

Iamgoofball commented on code/__HELPERS/icons.dm:
```diff
@@ -1086,23 +1086,9 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
-		remove_atom_colour(TEMPORARY_COLOUR_PRIORITY, GLOB.freon_color_matrix)
-		alpha += 25
-		obj_flags &= ~FROZEN
+	SEND_SIGNAL(src, COMSIG_OBJ_UNFREEZE)
```
This can be atomized

Iamgoofball commented on code/datums/components/overlay_lighting.dm:
```diff
@@ -111,14 +111,14 @@
 	RegisterSignal(parent, COMSIG_ATOM_UPDATE_LIGHT_FLAGS, .proc/on_light_flags_change)
 	RegisterSignal(parent, COMSIG_ATOM_USED_IN_CRAFT, .proc/on_parent_crafted)
 	RegisterSignal(parent, COMSIG_LIGHT_EATER_QUEUE, .proc/on_light_eater)
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, .proc/on_parent_moved)
```
Is there any way we can condense this block of RegisterSignal()s?

Mothblocks replied:
not really no

Kylerace replied:
not any sensible way no

Iamgoofball commented on code/game/atoms_movable.dm:
```diff
@@ -917,12 +928,13 @@
 	. = FALSE
-	move_stacks++
 	var/atom/oldloc = loc
+	var/is_multi_tile = bound_width > 32 || bound_height > 32
```
Magic numbers, we've got a define for the icon size we're using, please use that instead in the hypothetical future that we change our default icon size.

Kylerace replied:
done'd

Iamgoofball commented on code/game/turfs/open/space/space.dm:
```diff
@@ -22,6 +22,8 @@
 	bullet_bounce_sound = null
 	vis_flags = VIS_INHERIT_ID //when this be added to vis_contents of something it be associated with something on clicking, important for visualisation of turf in openspace and interraction with openspace that show you turf.
 
+	force_no_gravity = TRUE
```
This might make breaches more lethal than previously. Are you sure about this?

Kylerace replied:
what?

Kylerace replied:
this is a replacement for has_gravity() behavior where it typechecks the loc to see if its a space turf so it can return no gravity (unless the object has forced gravity), now it just checks for this var so any turf can do this.

Iamgoofball commented on code/modules/industrial_lift/crossing_signal.dm:
```diff
@@ -103,37 +103,43 @@
+	var/obj/structure/industrial_lift/tram/tram_part = tram.return_closest_platform_to(src)
+
+	if(QDELETED(tram_part))
+		set_signal_state(XING_STATE_GREEN, force = !is_operational)
```
We should probably convert this to signals and de-hardcode this, that way more things can listen for tram status.

Kylerace replied:
im not sure what you mean by this. what part should be converted to signals? the tram proximity check? if so sure, not in this pr but you could register to the tram entering turfs within a certain proximity and use that to determine the current state. not sure what you mean by de-hardcode though

Iamgoofball commented on code/modules/industrial_lift/industrial_lift.dm:
```diff
@@ -0,0 +1,753 @@
+	obj_flags = CAN_BE_HIT | BLOCK_Z_OUT_DOWN
+	appearance_flags = PIXEL_SCALE|KEEP_TOGETHER //no TILE_BOUND since we're potentially multitile
+
+	var/id = null //ONLY SET THIS TO ONE OF THE LIFT'S PARTS. THEY'RE CONNECTED! ONLY ONE NEEDS THE SIGNAL!
```
autodoc this

Kylerace replied:
this var doesnt appear to be used anymore so ill just remove it

Iamgoofball commented on code/modules/industrial_lift/lift_master.dm:
```diff
@@ -0,0 +1,395 @@
+
+	for(var/z in min_z to max_z)
+		//This must be safe way to border tile to tile move of bordered platforms, that excludes platform overlapping.
+		if( going & WEST )
```
weird spacing on all these ``if(going & DIR)`` checks

Kylerace replied:
i think that was how it was before, gone now

Iamgoofball commented on code/modules/industrial_lift/tram_override_objects.dm:
```diff
@@ -0,0 +1,44 @@
+ * the tram took a third of the tick per movement when its just carrying its default mapped in objects
+ */
+
+/obj/structure/grille/tram
```
these defines aren't needed

Kylerace replied:
i know i just think its clearer

Kylerace replied:
maybe its just me but defining type and proc in the same line is possible via behavior that has fucked me over in the past when i mispell a typepath and so it compiles an important proc to that newly defined type

Iamgoofball commented on code/modules/industrial_lift/tram_override_objects.dm:
```diff
@@ -0,0 +1,44 @@
+
+/obj/structure/grille/tram
+
+/obj/structure/grille/tram/Initialize(mapload)
```
because this defines it

Iamgoofball reviewed (changes requested):
good PR, just a few nitpicks

Kylerace committed:
addresses goofball reviews

Kylerace committed:
merges master and fixes conflicts

Kylerace committed:
removes id var and updates comment for has_gravity

Kylerace committed:
fixes a map error from removing the id variable

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into fasterer_tram

Kylerace committed:
uses the lift_id landmark instead of a var edit that doesnt work

LemonInTheDark commented:
swear I'll review this at some point

Kylerace committed:
removes moved_us argument for Moved() because its unused

Kylerace committed:
makes playsound() not use orthogonal_range_search() because its slower

turns out iterating a list of up to 100 people and calling get_dist() on each is faster than going through the spatial grid to find some less people and still iterating to call get_dist() on each of them. and the difference between the two versions of this part of this process is only a small part of the overall cost anyways

Kylerace commented on code/modules/mob/living/living_movement.dm:
```diff
@@ -28,9 +28,12 @@
 	if(isopenturf(T))
-		add_or_update_variable_movespeed_modifier(/datum/movespeed_modifier/turf_slowdown, multiplicative_slowdown = T.slowdown)
-	else
+		if(T.slowdown != current_turf_slowdown)
```
the direct movespeed modifier setters are surprisingly costly but since 99% of the time theres no movespeed modification from turfs and 99.999% of the time the movespeed modifier of the last turf is the same as the new turf, we can save almost the entire cost of this proc by just only updating it when its different

Kylerace commented on code/modules/mob/mob_movement.dm:
```diff
@@ -118,7 +118,9 @@
-	mob.set_glide_size(DELAY_TO_GLIDE_SIZE(add_delay * ( (NSCOMPONENT(direct) && EWCOMPONENT(direct)) ? SQRT_2 : 1 ) )) // set it now in case of pulled objects
+	var/new_glide_size = DELAY_TO_GLIDE_SIZE(add_delay * ( (NSCOMPONENT(direct) && EWCOMPONENT(direct)) ? SQRT_2 : 1 ) )
+	if(mob.glide_size != new_glide_size)
+		mob.set_glide_size(new_glide_size) // set it now in case of pulled objects
```
just to save on proc call overhead

LemonInTheDark replied:
don't trust this, see previous

Kylerace commented on code/modules/mob/mob_movement.dm:
```diff
@@ -358,10 +365,12 @@
 	var/speed_change = max(0, has_gravity - STANDARD_GRAVITY)
-	if(!speed_change)
+	if(!speed_change && gravity_slowdown)
 		remove_movespeed_modifier(/datum/movespeed_modifier/gravity)
```
another "only do work if anything has changed" optimization that should have been there from the start

Kylerace reviewed (commented):
more change annotations! hopefully this helps reviewers

Kylerace commented:
@LemonInTheDark went through and annotated more of my changes to hopefully try to make the different parts for each general movement optimization clearer in what their intent is.

LemonInTheDark commented on code/__HELPERS/icons.dm:
```diff
@@ -1086,23 +1086,9 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
-	if(obj_flags & FROZEN)
-		name = replacetext(name, "frozen ", "")
-		remove_atom_colour(TEMPORARY_COLOUR_PRIORITY, GLOB.freon_color_matrix)
-		alpha += 25
-		obj_flags &= ~FROZEN
+	SEND_SIGNAL(src, COMSIG_OBJ_UNFREEZE)
```
comment is wrong, also why does this proc exist

Kylerace replied:
only used by wet_floor component's process() it seems. wouldnt lose anything sin-wise in the current state by just making the call into the signal send but "unfreeze this obj" is relatively generic, if its ever used somewhere else we'd be doing the copypasted signal thing we all know and love

LemonInTheDark commented on code/datums/elements/footstep.dm:
```diff
@@ -134,21 +134,31 @@
 
-	play_fov_effect(source, 5, "footstep", direction, ignore_self = TRUE)
+	//cache for sanic speed (lists are references anyways)
+	var/static/list/footstep_sounds = GLOB.footstep
```
really? that bad?

Kylerace replied:
in retrospect im pretty sure this is almost useless since it now has to go from accessing a global datum and then a field of that datum to just accessing a global field. since its used three times it would be faster to make it a local reference

LemonInTheDark replied:
the fuck

LemonInTheDark commented on code/datums/elements/forced_gravity.dm:
```diff
@@ -1,16 +1,18 @@
+	var/ignore_turf_gravity
 
-/datum/element/forced_gravity/Attach(datum/target, gravity=1, ignore_space=FALSE)
+/datum/element/forced_gravity/Attach(datum/target, gravity=1, ignore_turf_gravity = FALSE)
```
weird inconsistent syntax

Kylerace replied:
true

LemonInTheDark commented on code/datums/elements/frozen.dm:
```diff
@@ -0,0 +1,47 @@
+	obj_source.alpha += 25
+	. = ..()
+
+/datum/element/frozen/proc/on_unfrozen(datum/source)
```
on unfreeze?

Kylerace replied:
https://www.youtube.com/watch?v=p2WzkgkufyE

LemonInTheDark commented on code/datums/elements/frozen.dm:
```diff
@@ -0,0 +1,47 @@
+	obj_target.visible_message(span_danger("[obj_target] shatters into a million pieces!"))
+	qdel(obj_target)
+
+/datum/element/frozen/proc/on_moved(datum/target)
+	SIGNAL_HANDLER
+	var/obj/obj_target = target
+	if(!isopenturf(obj_target.loc))
+		return
+
+	var/turf/open/turf_loc = obj_target.loc
+	if(turf_loc.air?.temperature >= T0C)//unfreezes target
+		Detach(target)
```
was this really only on moved

Kylerace replied:
technically it was on /turf/open/Entered() but it proc'd for all movable movement despite only applying to a tiny fraction of movables

LemonInTheDark replied:
hhh

LemonInTheDark commented on code/game/area/areas.dm:
```diff
@@ -403,7 +403,8 @@ GLOBAL_LIST_EMPTY(teleportlocs)
 	set waitfor = FALSE
 	SEND_SIGNAL(src, COMSIG_AREA_ENTERED, arrived, old_area)
-	if(!LAZYACCESS(arrived.important_recursive_contents, RECURSIVE_CONTENTS_AREA_SENSITIVE))
+
+	if(!arrived.important_recursive_contents?[RECURSIVE_CONTENTS_AREA_SENSITIVE])
```
why?

Kylerace replied:
lazyaccess is dumb and is designed for safety in a way thats arguably good in some contexts and completely useless in others while slowing down runtime in all contexts. since we're not checking a numerical index part of lazyaccess is redundant

LemonInTheDark replied:
is this code hot enough that avoiding the isnum is worth it? how large is the difference

Kylerace replied:
![Screenshot_2085](https://user-images.githubusercontent.com/15794172/175843657-ea80d09b-1bfd-4681-8e33-3410d1831ebc.png)
about 20% faster for this case not including the offsets from both being datum vars or ~90 nanoseconds per iteration. in that time, a photon of light could travel a whole 90 feet in a vacuum. a whole 90 feet! over the few tens of thousands of times this proc is called over the course of a round that photon could travel a few hundred miles. what amazing savings!

in seriousness i still dont like lazyaccess because it provides no actual value when you know youre reading an associative value and very debatable value when youre reading a value at a specific numerical index (shouldnt you know whether the list is that big, and if it isnt shouldnt you get an error?). 

the lazy macros should just be list getters and setters that are safe regardless of whether the list is null or not, and specifically for setters also null the list if totally empty

LemonInTheDark commented on code/game/atoms.dm:
```diff
@@ -1792,39 +1794,28 @@
+		if(!gravity_turf)//no gravity in nullspace
+			return 0
+
+	//the list isnt created every time as this proc is very hot, its only accessed if anything is actually listening to the signal too
+	var/static/list/forced_gravity = list()
+	if(SEND_SIGNAL(src, COMSIG_ATOM_HAS_GRAVITY, gravity_turf, forced_gravity))
+		if(!length(forced_gravity))
+			SEND_SIGNAL(gravity_turf, COMSIG_TURF_HAS_GRAVITY, src, forced_gravity)
```
won't this list persist then?

Kylerace replied:
yes, but its guaranteed to be cleared if the SEND_SIGNAL returns a truthy value

LemonInTheDark commented on code/game/atoms.dm:
```diff
@@ -1792,39 +1794,28 @@
-		// There's a gravity generator on our z level
-		if(GLOB.gravity_generators["[gravity_turf.z]"])
-			var/max_grav = 0
-			for(var/obj/machinery/gravity_generator/main/main_grav_gen as anything in GLOB.gravity_generators["[gravity_turf.z]"])
```
I don't understand this pattern of moving z level values onto lists on the subsystem

LemonInTheDark commented on code/game/atoms.dm:
```diff
@@ -1792,39 +1794,28 @@
-				max_grav = max(main_grav_gen.setting,max_grav)
-			return max_grav
-	return SSmapping.level_trait(gravity_turf.z, ZTRAIT_GRAVITY)
+	return !gravity_turf.force_no_gravity && (SSmapping.gravity_by_z_level["[gravity_turf.z]"] || turf_area.has_gravity)
```
wouldn't it be better to check the area first, then the z level?

Kylerace replied:
yes it would

LemonInTheDark commented on code/game/atoms_movable.dm:
```diff
@@ -631,12 +625,12 @@
 	//glide_size strangely enough can change mid movement animation and update correctly while the animation is playing
 	//This means that if you don't override it late like this, it will just be set back by the movement update that's called when you move turfs.
-	if(glide_size_override)
+	if(glide_size_override && glide_size != glide_size_override)
 		set_glide_size(glide_size_override)
```
not always valid, glide size being set also effects things sitting on us

LemonInTheDark commented on code/game/atoms_movable.dm:
```diff
@@ -661,26 +655,25 @@
-	if(move_stacks < 0)
-		stack_trace("move_stacks is negative in Moved()!")
-		move_stacks = 0 //setting it to 0 so that we dont get every movable with negative move_stacks runtiming on every movement
+	SEND_SIGNAL(src, COMSIG_MOVABLE_MOVED, old_loc, movement_dir, forced, old_locs, momentum_change)
```
not sure I trust this

LemonInTheDark commented on code/game/turfs/open/_open.dm:
```diff
@@ -187,10 +188,9 @@
-			continue
-		if(!(I.obj_flags & FROZEN))
-			I.make_frozen_visual()
+		if(!HAS_TRAIT(I, TRAIT_FROZEN) && !(I.obj_flags & FREEZE_PROOF))
```
an I dumb? where is TRAIT_FROZEN used?

Kylerace replied:
freeze_turf() and /obj/item/freeze_cube/proc/freeze()

Kylerace replied:
wait i see what you mean, i never add the trait! not very good, i explicitly remember encountering a bug with the frozen element being added multiple times in the tm and i thought i fixed it there but it repeated after merge

LemonInTheDark commented on code/game/turfs/turf.dm:
```diff
@@ -363,8 +365,6 @@ GLOBAL_LIST_EMPTY(station_turfs)
 	var/canPassSelf = CanPass(mover, get_dir(src, mover))
 	if(canPassSelf || (mover.movement_type & PHASING))
 		for(var/atom/movable/thing as anything in contents)
-			if(QDELETED(mover))
-				return FALSE //We were deleted, do not attempt to proceed with movement.
```
🤨

Kylerace replied:
oh this is slightly incorrect, the only impure things that can delete mover in this loop is the Bump() called iff mover is phasing or the Cross() call if some weirdo eventually creates an override that can delete mover from there

LemonInTheDark commented on code/game/turfs/turf.dm:
```diff
@@ -385,14 +385,6 @@ GLOBAL_LIST_EMPTY(station_turfs)
-	//melting
-	if(isobj(arrived) && air && air.temperature > T0C)
-		var/obj/O = arrived
-		if(O.obj_flags & FROZEN)
```
this is so dumb bro

LemonInTheDark commented on code/modules/lighting/lighting_source.dm:
```diff
@@ -74,6 +74,35 @@
 
+///add this light source to new_atom_host's light_sources list. updating movement registrations as needed
+/datum/light_source/proc/add_to_light_sources(atom/new_atom_host)
+	if(QDELETED(new_atom_host))
+		return FALSE
```
do we need this check?

Kylerace replied:
yes unless youd rather it error i suppose

LemonInTheDark commented on code/modules/mob/mob_movement.dm:
```diff
@@ -137,10 +139,15 @@
-		mob.set_glide_size(DELAY_TO_GLIDE_SIZE(add_delay))
+		after_glide = DELAY_TO_GLIDE_SIZE(add_delay)
+
+	if(after_glide != mob.glide_size)
+		mob.set_glide_size(after_glide)
```
see previous

LemonInTheDark commented on code/modules/shuttle/shuttle.dm:
```diff
@@ -68,7 +68,7 @@
 		return QDEL_HINT_LETMELIVE
 
 /obj/docking_port/has_gravity(turf/T)
-	return FALSE
+	return TRUE
```
hm?

Kylerace replied:
somewhere during development something made docking ports start drifting away roundstart and it fucked up every shuttle. somewhat hacky fix to that problem but other than that i dont disagree with this change in a vacuum. i added a real fix for this after i traced it to changes in abstract_move()

Kylerace commented on code/__DEFINES/industrial_lift.dm:
```diff
@@ -0,0 +1,20 @@
+///the lift's controls are currently unlocked so user's can direct it
+#define LIFT_PLATFORM_UNLOCKED 0
+
+//lift_id's
+///basic lift_id, goes up and down
+#define BASIC_LIFT_ID "base"
+///tram lift_id, goes left and right or north and south. maybe one day be able to turn and go up/down as well
+#define TRAM_LIFT_ID "tram"
+///debug lift_id
+#define DEBUG_LIFT_ID "debug"
```
groups are a better name i think

Kylerace commented on code/__DEFINES/industrial_lift.dm:
```diff
@@ -0,0 +1,20 @@
+#define DEBUG_LIFT_ID "debug"
+
+
+//specific_lift_id's
+///the specific_lift_id of the main station tram landmark for tramstation that spawns roundstart.
+#define MAIN_STATION_TRAM "main station tram"
+///the specific_lift_id of the tram on the hilbert research station
+#define HILBERT_TRAM "tram_hilbert"
```
id's linked to one lift in the world

Kylerace commented on code/_onclick/click.dm:
```diff
@@ -213,11 +213,13 @@
 		for(var/atom/target in checking)  // will filter out nulls
 			if(closed[target] || isarea(target))  // avoid infinity situations
 				continue
-			closed[target] = TRUE
+
 			if(isturf(target) || isturf(target.loc) || (target in direct_access) || (ismovable(target) && target.flags_1 & IS_ONTOP_1)) //Directly accessible atoms
 				if(Adjacent(target) || (tool && CheckToolReach(src, target, tool.reach))) //Adjacent or reaching attacks
 					return TRUE
 
+			closed[target] = TRUE
+
```
microest of micro opts since closed isnt accessed at all inside those if statements but they can return so by putting it after the check you have a chance of not needing to do the associative write

LemonInTheDark replied:
dirty, disgusting, impressive

Kylerace commented on code/controllers/subsystem/mapping.dm:
```diff
@@ -40,14 +40,21 @@ SUBSYSTEM_DEF(mapping)
 	var/space_levels_so_far = 0
-	///list of all the z level datums created representing the z levels in the world
-	var/list/z_list
+	///list of all z level datums in the order of their z (z level 1 is at index 1, etc.)
+	var/list/datum/space_level/z_list
+	///list of all z level indices that form multiz connections and whether theyre linked up or down
+	///list of lists, inner lists are of the form: list("up or down link direction" = TRUE)
+	var/list/multiz_levels = list()
```
1. should have put in a period at the end of the first line for the second var but its: line 1 being purpose of the var and line 2 being structure of the list.
2. multiz data structures are generally the domain of SSmapping

Kylerace commented on code/controllers/subsystem/mapping.dm:
```diff
@@ -40,14 +40,21 @@ SUBSYSTEM_DEF(mapping)
 	/// True when in the process of adding a new Z-level, global locking
 	var/adding_new_zlevel = FALSE
 
+	///shows the default gravity value for each z level. recalculated when gravity generators change.
+	///associative list of the form: list("[z level num]" = max generator gravity in that z level OR the gravity level trait)
+	var/list/gravity_by_z_level = list()
+
```
its part of the optimizations to has_gravity() so we dont have to check all the gravity generators in a z level and then default z level gravity when the resulting value is 1. pertaining to a specific z level and 2. changes only when either there is a change in that z levels gravity generators or its gravity level trait

Kylerace commented on code/controllers/subsystem/spatial_gridmap.dm:
```diff
@@ -171,53 +171,6 @@ SUBSYSTEM_DEF(spatial_grid)
-		assigned_atom = atoms_that_need_ears[current_ear_index]
-
-		turf_loc = get_turf(assigned_atom)
-		if(!turf_loc)
-			continue
-
-		current_ear = pregenerated_oranges_ears[current_ear_index]
-
-		if(turf_loc.assigned_oranges_ear)
-			turf_loc.assigned_oranges_ear.references += assigned_atom
-			continue //if theres already an oranges_ear mob at assigned_movable's turf we give assigned_movable to it instead and dont allocate ourselves
-
-		current_ear.references += assigned_atom
-
```
mostly. theres a removal in old lines 281-284 and an addition in new line 410

Kylerace commented on code/datums/components/bloodysoles.dm:
```diff
@@ -235,7 +258,6 @@
 		bloody_feet = mutable_appearance('icons/effects/blood.dmi', "shoeblood", SHOES_LAYER)
 
 	RegisterSignal(parent, COMSIG_COMPONENT_CLEAN_ACT, .proc/on_clean)
```
because its easier and unlike movement it doesnt happen much

LemonInTheDark replied:
isn't it the same as movement?

Kylerace replied:
i guess in the sense that it only does anything if its dirty? i dont see much point in changing it though

LemonInTheDark replied:
it feels kinda messy to like uh, do stuff like that in such a limited space only for optimization, messes with how people reason about it if that makes sense

Kylerace commented on code/__HELPERS/icons.dm:
```diff
@@ -1086,23 +1086,9 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
 
 GLOBAL_LIST_INIT(freon_color_matrix, list("#2E5E69", "#60A2A8", "#A1AFB1", rgb(0,0,0)))
```
moved

Kylerace commented on code/datums/components/drift.dm:
```diff
@@ -114,7 +114,6 @@
 		return
 
 	var/atom/movable/movable_parent = parent
-	movable_parent.inertia_moving = FALSE
```
holdover from an attempt at something that would need its own pr, put back in

Kylerace commented on code/__DEFINES/obj_flags.dm:
```diff
@@ -9,7 +9,6 @@
 #define ON_BLUEPRINTS (1<<5)  //Are we visible on the station blueprints at roundstart?
 #define UNIQUE_RENAME (1<<6) // can you customize the description/name of the thing?
 #define USES_TGUI (1<<7) //put on things that use tgui on ui_interact instead of custom/old UI.
-#define FROZEN (1<<8)
```
done
